front/lib/api/assistant/generation.ts

```diff
@@ -255,6 +272,196 @@ export async function renderConversationForModel({
+        if (isRetrievalActionType(m.action)) {
+          messages.unshift(renderFunctionMessageForRetrievalAction(m.action));
+          messages.unshift(
+            renderAssistantFunctionCallMessageForRetrievalAction(m.action)
+          );
+        } else if (isDustAppRunActionType(m.action)) {
+          messages.unshift(
+            renderFunctionMessageForForDustAppRunAction(m.action)
+          );
+          messages.unshift(
+            renderAssistantFunctionCallMessageForDustAppRunAction(m.action)
+          );
+        } else if (isTablesQueryActionType(m.action)) {
+          messages.unshift(
+            renderFunctionMessageForForTablesQueryAction(m.action)
+          );
+          messages.unshift(
+            renderAssistantFunctionCallMessageForTablesQueryAction(m.action)
+          );
+        } else if (isProcessActionType(m.action)) {
+          messages.unshift(renderFunctionMessageForForProcessAction(m.action));
+          messages.unshift(
+            renderAssistantFunctionCallMessageForProcessAction(m.action)
+          );
+        } else {
+          assertNever(m.action);
+        }
+      }
+    } else if (isUserMessageType(m)) {
+      // Replace all `:mention[{name}]{.*}` with `@name`.
+      const content = m.content.replaceAll(
+        /:mention\[([^\]]+)\]\{[^}]+\}/g,
+        (_, name) => {
+          return `@${name}`;
+        }
+      );
+      messages.unshift({
+        role: "user" as const,
+        name: m.context.fullName || m.context.username,
+        content,
+      });
+    } else if (isContentFragmentType(m)) {
+      try {
+        const content = await getContentFragmentText({
+          workspaceId: conversation.owner.sId,
+          conversationId: conversation.sId,
+          messageId: m.sId,
+        });
+        messages.unshift({
+          role: "content_fragment",
+          name: `inject_${m.contentType}`,
+          content:
+            `TITLE: ${m.title}\n` +
+            `TYPE: ${m.contentType}${
+              m.contentType === "file_attachment" ? " (user provided)" : ""
+            }\n` +
+            `CONTENT:\n${content}`,
+        });
+      } catch (error) {
+        logger.error(
+          {
+            error,
+            workspaceId: conversation.owner.sId,
+            conversationId: conversation.sId,
+            messageId: m.sId,
+          },
+          "Failed to retrieve content fragment text"
+        );
+        return new Err(new Error("Failed to retrieve content fragment text"));
+      }
+    } else {
+      assertNever(m);
+    }
+  }
+
+  // Compute in parallel the token count for each message and the prompt.
+  const [messagesCountRes, promptCountRes] = await Promise.all([
+    Promise.all(
+      messages.map((m) => {
+        let text = `${m.role} ${"name" in m ? m.name : ""} ${m.content ?? ""}`;
+        if ("function_calls" in m) {
+          text += m.function_calls
+            .map((f) => `${f.function.name} ${f.function.arguments}`)
```

Comment:
Will be : `search_data_sources {"query":"Soupinou cute","relativeTimeFrame":"all","topK":32}`